#162163262 implement jwt authentication feature

Kasulejoseph opened the pull request:
**what does this PR do?**
Generate and return a token to a user which is used to access all authenticated endpoints
**Description of tasks to be completed**
- writing unit tests for this feature
- Allowing access to the following endpoints
`GET /api/user` 
`PUT /api/user`
`GET /api/profiles/:username`
`POST /api/profiles/:username/follow`
`DELETE /api/profiles/:username/follow`
`GET /api/articles`
`GET /api/articles/feed`
`POST /api/articles`
`PUT /api/articles/:slug`
`DELETE /api/articles/:slug`
`POST /api/articles/:slug/comments`
`GET /api/articles/:slug/comments`
`DELETE /api/articles/:slug/comments/:id`
`POST /api/articles/:slug/favorite`
`DELETE /api/articles/:slug/favorite`

**How this can be tested manually**
- After cloning the repo, install the requirements with `pip install -r requirements.txt`
- To run tests, run `./manage.py test`
- Start the server and try out the following endpoints
registration endpoint data format--> `POST api/users/`
```
{
	"user":{
			"email":"kasule@mail56859.com",
			"username":"kasule56859",
			"password":"kasule89990"
			}
}
```
Login data format --> `POST /api/users/login/`
```
`{
  "user":{
			 "email": "kasule@mail56859.com",
			"password":"kasule89990"
  }
}
```
expected output on registration 
```
{
    "user": {
        "email": "kasule@mail56859.com",
        "username": "kasule56859",
        "token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6OSwibmFtZSI6Imthc3VsZTU2ODU5IiwiZXhwIjoxNTQ0MDgyODgxfQ.2pZPEl9gNMCbftfqljOv9FYISHVOCeL104uEUlUH9-I"
    }
}
```
- User the token provided to access all other protected endpoints
example is the get user endpoint `GET /api/user/` 
<img width="965" alt="screen shot 2018-12-05 at 17 46 58" src="https://user-images.githubusercontent.com/32167860/49521290-2bf4cd00-f8b6-11e8-9028-efe464b61456.png">

**Screetshots**
Invalid token
<img width="955" alt="screen shot 2018-12-05 at 17 48 43" src="https://user-images.githubusercontent.com/32167860/49521235-11225880-f8b6-11e8-91ec-ada133741180.png">

**Relevant Story**
- #162163262

Commit:
feat(jwt auth): implement jwt authentication

return a token upon successful login
allow a user to access authenticated  endpoints with the generated token
move SECRET_KEY from settings.py to .env

[Delivers #162163262]

Kasulejoseph committed:
feat(jwt auth): implement jwt authentication

return a token upon successful login
allow a user to access authenticated  endpoints with the generated token
move SECRET_KEY from settings.py to .env

[Delivers #162163262]

Kasulejoseph committed:
feat(tests jwt): write tests for jwt auth

re-structure the test files
add tests for jwt authentication feature
raise coverage to 91%

[Delivers #162163262]

Kasulejoseph committed:
feat(tests jwt): write tests for jwt auth

re-structure the test files
add tests for jwt authentication feature
raise coverage to 91%

[Delivers #162163262]

bibangamba commented:
Hi @Kasulejoseph, this is DOPE work!! Some suggestions, you have a mistake in "what this PR do?". It should be, "What does this PR do?". Then there's also a bunch of typos in the body.

I've tested the app via its review app, attached just below your last commit (shown as View Deployment). For the tests, I just used the attached Travis links to check out the test coverage. I say this to show an alternative 'manual test'.

Everything seems to work well, just a suggestion on the Register/Sign up function. Can we have it create the authentication token on successful registration? Essentially it would work like we've logged the user in on signup. 
Maybe @collin5 can advise on this. I think it would be a good idea because when we are implementing the UI, we would not want the user to sign up, get confirmation that they are now registered, then ask them to sign in (although we could pass their password from signup as an argument to some function and then redirect them to log in endpoint... hmmm. I think I'll go with whatever the team or Collins advises).

Nta1e commented:
👏🏽. Good work on the Jwt, also on hiding the secret key in the environment.🤝
Just fix some typos in the description like `pip install -tr requirements.txt`

Kasulejoseph committed:
feat(jwt auth): write test for jwt auth

re-structure the test files
add tests for jwt authentication feature
raise coverage to 91%

[Delivers #162163262]

Kasulejoseph committed:
feat(jwt auth): write test for jwt auth

re-structure the test files
add tests for jwt authentication feature
raise coverage to 91%

[Delivers #162163262]

Kasulejoseph commented:
> Hi @Kasulejoseph, this is DOPE work!! Some suggestions, you have a mistake in "what this PR do?". It should be, "What does this PR do?". Then there's also a bunch of typos in the body.
> 
> I've tested the app via its review app, attached just below your last commit (shown as View Deployment). For the tests, I just used the attached Travis links to check out the test coverage. I say this to show an alternative 'manual test'.
> 
> Everything seems to work well, just a suggestion on the Register/Sign up function. Can we have it create the authentication token on successful registration? Essentially it would work like we've logged the user in on signup.
> Maybe @collin5 can advise on this. I think it would be a good idea because when we are implementing the UI, we would not want the user to sign up, get confirmation that they are now registered, then ask them to sign in (although we could pass their password from signup as an argument to some function and then redirect them to log in endpoint... hmmm. I think I'll go with whatever the team or Collins advises).

Thank you for the review @bibangamba, I have worked on the grammatical errors and about the token, after discussing with most of the teammates we agreed that the token will only be given on user login and not signup. thanks

Kasulejoseph commented:
> 👏🏽. Good work on the Jwt, also on hiding the secret key in the environment.🤝
> Just fix some typos in the description like `pip install -tr requirements.txt`

your welcome, i have fixed the typo error

Kasulejoseph committed:
feat(jwt auth): write test for jwt auth

re-structure the test files
add tests for jwt authentication feature
raise coverage to 91%

[Delivers #162163262]

collin5 commented on authors/apps/authentication/backends.py:
```diff
@@ -1,14 +1,51 @@
+        if header:
+            token = header.split()[1].decode('utf-8')
+            return token
+        return None
```
Raise an Exception if token is None

Kasulejoseph replied:
Hey, thanks for the review. However, I still have issues implementing the Exception feedback. The `jwtauthentication` class must either return a `token` to endpoints that are authenticated or `None` to endpoints that do not require authentication. In this logic when I raise an exception, the endpoints that do not require authentication fails. I welcome any suggestion you give me. thanks

collin5 commented on authors/apps/authentication/backends.py:
```diff
@@ -1,14 +1,51 @@
+        a user object and token are returned.
+        """
+        user_token = self.authenticate_header(request)
+        if not user_token:
```
This `if` block won't be required if above ☝️ feedback is implemented. You may/may-not need a try block instead.

collin5 commented on authors/apps/authentication/backends.py:
```diff
@@ -1,14 +1,51 @@
+        try:
+            payload = jwt.decode(user_token, settings.SECRET_KEY)
+        except jwt.InvalidTokenError:
+            invalid_error = 'Invalid token. please login again'
```
Just put this message in parenthesis of the Exception. Same goes for other exceptions below.

collin5 reviewed (changes requested):
Good work! Few stuff to resolve before we merge.

Kasulejoseph committed:
feat(jwt auth): implement feedback

raise an exception for empty token in the header

~:[Delivers #162163262]
